src/test/java/org/jenkinsci/plugins/scriptsecurity/sandbox/groovy/SandboxInterceptorTest.java

```diff
@@ -338,6 +338,27 @@ public static final class Dynamic extends GroovyObjectSupport {
         }
     }
 
+    @Test public void propertyMissingMethods() throws Exception {
```

Comment:
Right, my test might not fully reproduce the problem, especially when tested against old versions. My goal with the test was mainly to prevent future regressions.
To give some more context. Here's a Jenkins pipeline script that worked before https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 but fails afterwards:
```
class A {
    
    def values = [:]
    
    @NonCPS
    public void propertyMissing(String name, Object value) {
        values[name] = value
    }
}

def a = new A()

a.foo = 'bar'
echo "${a.values}"
```
Let's have a look at the call stack inside of A.propertyMissing():
```
	...
	at A.propertyMissing(WorkflowScript:7)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:98)
	at groovy.lang.MetaClassImpl.invokeMissingProperty(MetaClassImpl.java:894)
	at groovy.lang.MetaClassImpl.setProperty(MetaClassImpl.java:2751)
	at groovy.lang.MetaClassImpl.setProperty(MetaClassImpl.java:3785)
	at A.setProperty(WorkflowScript)
	at org.codehaus.groovy.runtime.InvokerHelper.setProperty(InvokerHelper.java:213)
	at org.codehaus.groovy.runtime.ScriptBytecodeAdapter.setProperty(ScriptBytecodeAdapter.java:497)
	at org.kohsuke.groovy.sandbox.impl.Checker$8.call(Checker.java:412)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onSetProperty(GroovyInterceptor.java:84)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onSetProperty(SandboxInterceptor.java:213)
	at org.kohsuke.groovy.sandbox.impl.Checker$8.call(Checker.java:409)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedSetProperty(Checker.java:416)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.setProperty(SandboxInvoker.java:33)
	at com.cloudbees.groovy.cps.impl.PropertyAccessBlock.rawSet(PropertyAccessBlock.java:24)
	at WorkflowScript.run(WorkflowScript:13)
	...
```
We can see that we end up in SandboxInterceptor.onSetProperty after the call to a.foo on line 13. Before the fix, this was fine and we continue to the super implementation in GroovyInterceptor.onSetProperty, we continue through the synthetic method A.setProperty, which causes us to end up in groovy.lang.MetaClassImpl.setProperty and with the help of groovy.lang.MetaClassImpl.invokeMissingProperty we eventually end up in A.propertyMissing on line 7.

After https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 the super call to GroovyInterceptor.onSetProperty is prohibited since A.setProperty is synthetic.

Now that I think of it I'm guessing that my test fails with the old version since it is a regular java class and lacking the synthetic setProperty method?

Reply:
Ok, thanks, I will try to take a deeper look later this week.

Reply:
Did some investigation. The following test seems to reproduce the problem, both in the old version and on master:
```
    @Test public void propertyMissingGroovyClass() throws Exception {
        String script =
                "class A {\n" +
                        "    def values = [:]\n" +
                        "    def propertyMissing(n) { values[n] }\n" +
                        "    def propertyMissing(n, v) { values[n] = v }\n" +
                        "}\n" +
                        "def a = new A()\n" +
                        "a.prop = 'val'\n" +
                        "a.prop";
        assertEvaluate(new GenericWhitelist(), "val", script);
    }
```
But, it turns out that our script approval list in Jenkins allowed 'method groovy.lang.GroovyObject setProperty java.lang.String java.lang.Object' (which is blacklisted), which made it work with the old version (in our defense we added it to our approval list, way, way back before it was blacklisted). 

So without my fix, prior to https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 the script above fails with:
```org.jenkinsci.plugins.scriptsecurity.sandbox.RejectedAccessException: Scripts not permitted to use method groovy.lang.GroovyObject setProperty java.lang.String java.lang.Object (A.prop)```
while afterwards it fails with:
```groovy.lang.MissingPropertyException: No such field found: field A prop```

Here's a test which more closely reproduces our setup:
```
    @Test public void propertyMissingGroovyClass() throws Exception {
        String script =
                "class A {\n" +
                        "    def values = [:]\n" +
                        "    def propertyMissing(n) { values[n] }\n" +
                        "    def propertyMissing(n, v) { values[n] = v }\n" +
                        "}\n" +
                        "def a = new A()\n" +
                        "a.prop = 'val'\n" +
                        "a.prop";
        String getP = "method groovy.lang.GroovyObject getProperty java.lang.String";
        String setP = "method groovy.lang.GroovyObject setProperty java.lang.String java.lang.Object";
        assertEvaluate(new ProxyWhitelist(new GenericWhitelist(), new StaticWhitelist(getP, setP)), "val", script);
    }
```
I.e. approve getProperty and setProperty (which is very unwise, very well aware of that). It passes prior to https://github.com/jenkinsci/script-security-plugin/commit/85d16bd851b355458f6396dbc4f92c3e0eebcd72 but fails afterwards with same error as above.

So, apologies, this issue was not as simple as first thought and mostly down to our negligence. I should also have tested on a clean Jenkins instance... But on the flip side, if this change is integrated, then we can remove the approval of the blacklisted metod.

Reply:
Thanks for the investigation! Just as a quick disclaimer, if while looking into this further you find something that looks like a security issue in the current code please file it on https://issues.jenkins.io/ as a **SECURITY** type issue instead of commenting on this PR so we can continue the discussion in private.

 In general there are a few reasons I am apprehensive about this kind of change.
1. Primarily, because we have to be extremely careful to not introduce security vulnerabilities. For example, with the patch as-is, if you define a Groovy class with the `propertyMissing` method pair, then you can bypass the sandbox with code like `a.metaClass = b.metaClass`, because it thinks Groovy will call `propertyMissing` when in fact it will call `getMetaClass` and `setMetaClass`. Perhaps it could be made safe by only checking for `propertyMissing` methods at the very end of the `SandboxInteceptor` methods when `rejector == null`, but I will have to think about it.
2. Usage of Groovy metaprogramming in sandboxed scripts is very often insecure and/or buggy, especially once you add in the CPS transformation in Pipelines, so I am hesitant to encourage this kind of scripting even if we can find a way to make it superficially safe. My recommendation for these cases is to use unsandboxed global Pipeline libraries instead to encapsulate the metaprogramming behavior or to define the classes that use metaprogramming, in which case the problem is avoided.
3. As a distant third concern, Groovy not only supports `propertyMissing`, but also `methodMissing`, `$static_propertyMissing`, and `$static_methodMissing`, which all have similar semantics. If we are going to support one, should we support others? Do we need to consider undesirable interactions if we only support one or another? Is it even possible to implement all of them safely? I am not sure.

Reply:
Thank you for your answer and looking into this. It has become painfully obvious to me that I apparently don't know enough of the inner workings of Groovy to address this.
> Perhaps it could be made safe by only checking for propertyMissing methods at the very end of the SandboxInteceptor methods when rejector == null, but I will have to think about it.

As far as I can tell, from the code, moving to the end wouldn't resolve the security problem that you highlighted. If I understand the logic and call sequence, this class intercepts the calls and then continues the call chain by calling super if the method/field is deemed safe. It is the up to the underlying implementation to decide what happens, and in this case the call will end up in the synthetic method setProperty (or getProperty), which allows for bad things to happen.
I guess the best solution would be to redirect the call to the actual propertyMissing method, but I'm guessing that that isn't possible. Another solution would be to use StaticWhitelist.isBlacklisted() to detect all bad calls and reject them. Not ideal since it would block legit calls to propertyMissing...

> Usage of Groovy metaprogramming in sandboxed scripts is very often insecure and/or buggy, especially once you add in the CPS transformation in Pipelines, so I am hesitant to encourage this kind of scripting even if we can find a way to make it superficially safe. My recommendation for these cases is to use unsandboxed global Pipeline libraries instead to encapsulate the metaprogramming behavior or to define the classes that use metaprogramming, in which case the problem is avoided.

Fully aware of this, unfortunately, in order to make user friendly frameworks, we ended up using this in interface layer between the user and our global pipeline. The thing with Jenkins and especially Jenkins pipeline is that it allows us to build big and complex systems/libraries that resolves our complex build, product, environment. This is the main differentiator for us that makes Jenkins stand out in the crowd of other CI systems. Short of writing our own CI system there is simply no other system that is flexible enough to accommodate the complex situations that big companies sometimes end up in.

But, we have realized that our solution isn't ideal, and looking for ways around it, on possible solution is this change, but we will also see if we can rearrange things in our pipeline so that we can use getProperty and setProperty instead (last time I tried, years ago, there were numerous problems with that approach).

> As a distant third concern, Groovy not only supports propertyMissing, but also methodMissing, $static_propertyMissing, and $static_methodMissing, which all have similar semantics. If we are going to support one, should we support others? Do we need to consider undesirable interactions if we only support one or another? Is it even possible to implement all of them safely? I am not sure.

Indeed it would be great to have support for these as well, but I can imagine that you, as a maintainer, rather spend your time on other things. It's a rarely used feature with big security concerns (as made obvious with my naive implementation). For me, as a user, I'm happy with every little improvement that is made, even if it is on a best effort level with partial support.

Anyways, I fully understand your scepticism towards my change, its a change in extremely security sensitive code.

Reply:
Generally speaking, we cannot guarantee support for metaprogramming or other tricky Groovy constructs in Pipeline script. You should be able to build a sophisticated library that still uses a surface syntax based on a mixture of static (Java-like) code and commonplace Groovy features such as closures and dynamic typing, which will be easier to reason about (if not always as pretty) and unlikely to be broken by subtle changes in `script-security` or `workflow-cps` plugins.

Reply:
Fully aware of that, I've seen similar comments from you before Jesse. Yet here we are, we tried really hard to not be dependent on metaprogramming, but after several iteration we still ended up being dependent upon it. Simply can't achieve the flexibility and ease of use without it.. And for years this has worked well, this is the first time that we have run into problems. With all of this said, we only have ourselves to blame for this regression, and we were obviously depending on a security hole that had to be mended. Additionally, while it initially wasn't (or still isn't?) stated in any official documentation that things like this should be avoided, once https://www.jenkins.io/blog/2017/02/01/pipeline-scalability-best-practice/ and similar posts started to appear, we should have starting to reconsider our decisions. Unfortunately by then we were stuck with something that works, extremely flexible and loved by our users.

Anyways, enough ranting about why our setup sucks, I created this PR with the hope that this was something simple and a way to get official support for this. Unfortunately I greatly underestimated how complex Groovy and sandbox interception is. I'm willing to help out to find a solution here, fully aware though that any code that I bring to the table probably takes longer time to review than it would take for someone with insights into the code to just fix it...